src/index.c

```diff
@@ -2301,33 +2329,48 @@ static size_t read_entry(
 	} else
 		path_ptr = (const char *) buffer + offsetof(struct entry_short, path);
 
-	path_length = entry.flags & GIT_IDXENTRY_NAMEMASK;
+	if (!compressed) {
+		path_length = entry.flags & GIT_IDXENTRY_NAMEMASK;
 
-	/* if this is a very long string, we must find its
-	 * real length without overflowing */
-	if (path_length == 0xFFF) {
-		const char *path_end;
+		/* if this is a very long string, we must find its
+		 * real length without overflowing */
+		if (path_length == 0xFFF) {
+			const char *path_end;
 
-		path_end = memchr(path_ptr, '\0', buffer_size);
-		if (path_end == NULL)
-			return 0;
+			path_end = memchr(path_ptr, '\0', buffer_size);
+			if (path_end == NULL)
+				return 0;
 
-		path_length = path_end - path_ptr;
-	}
+			path_length = path_end - path_ptr;
+		}
 
-	if (entry.flags & GIT_IDXENTRY_EXTENDED)
-		entry_size = long_entry_size(path_length);
-	else
-		entry_size = short_entry_size(path_length);
+		if (entry.flags & GIT_IDXENTRY_EXTENDED)
+			entry_size = long_entry_size(path_length);
+		else
+			entry_size = short_entry_size(path_length);
 
-	if (INDEX_FOOTER_SIZE + entry_size > buffer_size)
-		return 0;
+		if (INDEX_FOOTER_SIZE + entry_size > buffer_size)
+			return 0;
 
-	entry.path = (char *)path_ptr;
+		entry.path = (char *)path_ptr;
+	} else {
+		size_t shared = decode_varint((const unsigned char **) &path_ptr);
+		size_t len = strlen(path_ptr);
+		size_t last_len = strlen(*last);
 
-	if (index_entry_dup(out, index, &entry) < 0)
+		tmp_path = git__malloc(shared + len + 1);
```

Comment:
Also, `decode_varint` returns a `uintmax_t`, which could be larger than a `size_t`.  Since we are probably using these varints to hold array sizes / indexes, perhaps having encode/decode use `size_t`s makes more sense.

Independent of return types, it looks like this arithmetic could overflow if somebody stuffs a varint value that is (say) `SIZE_MAX - 1`  It looks like we would then write outside of `tmp_path` and corrupt the heap.